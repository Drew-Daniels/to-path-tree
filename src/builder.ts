import { DEFAULT_SEP } from './constants'
import type { NodeItem, ParseResults, TreeNode } from './types'
import { genRoot, parsePath } from './utils'

export class PathTreeBuilder<T extends Record<string, unknown>> {
  #root: TreeNode<T> = genRoot()
  #mapping = new Map<string, ParseResults<T>>()

  #sep
  #getData

  /**
   * @param options
   * @param options.sep - default `/`
   * @param options.getData - function to get data from node
   */
  constructor(options?: {
    sep?: string
    getData?: (node: NodeItem<T>) => T
  }) {
    const { sep = DEFAULT_SEP, getData } = options ?? {}

    this.#sep = sep
    this.#getData = getData
  }

  addPath(path: string) {
    const nodeItem = parsePath(path, this.#sep, this.#root, this.#getData)
    this.#mapping.set(path, nodeItem)
  }

  removePath(path: string) {
    const nodeItem = this.#mapping.get(path)
    if (!nodeItem)
      return
    const { node, type } = nodeItem
    if (type === 'file')
      node.parent?.items.splice(node.parent.items.indexOf(node), 1)
    else
      Reflect.deleteProperty(node.parent!.subDirectory!, node.name)
  }

  /**
   *
   * @param path path to get node
   * @param _isFile pass isFile for better type inference
   */
  getNode<const IF extends boolean>(path: string, _isFile: IF = true as IF) {
    const node = this.#mapping.get(path)
    if (!node)
      return null
    return node.node as IF extends true ? NodeItem<T> : TreeNode<T>
  }

  getItems(path: string) {
    const node = this.#mapping.get(path)
    if (!node)
      return []
    return node.type === 'file'
      ? []
      : node.node.items
  }

  get tree() {
    return this.#root
  }
}
