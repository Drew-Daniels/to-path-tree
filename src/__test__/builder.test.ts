import { ROOT_NAME } from '..'
import { PathTreeBuilder } from '@/builder'

it('simple builder', () => {
  const builder = new PathTreeBuilder()
  builder.addPath('/a/')
  builder.addPath('/a/b/')
  builder.addPath('/a/b/index.ts')
  builder.addPath('/a/b/foo.ts')
  expect(builder.tree).toMatchSnapshot('simple-builder add path')
  builder.removePath('/a/b/')
  builder.removePath('/a/b/foo.ts')
  builder.removePath('/a/b/baz.ts') // remove non-exist path
  expect(builder.tree).toMatchSnapshot('simple-builder remove path')
})

it('get items', () => {
  const builder = new PathTreeBuilder()
  builder.addPath('/a/')
  builder.addPath('/a/index.ts')
  expect(builder.tree).toMatchInlineSnapshot(`
    {
      "items": [],
      "name": "root",
      "path": "/",
      "relativePath": "/",
      "relativePathName": "",
      "subDirectory": {
        "a": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "/a/index.ts",
            },
          ],
          "name": "a",
          "parent": [Circular],
          "path": "/a/",
          "relativePath": "a/",
          "relativePathName": "a",
          "subDirectory": {},
        },
      },
    }
  `)
  expect(builder.getItems('/a/')).toMatchInlineSnapshot(`
    [
      {
        "data": undefined,
        "ext": "ts",
        "file": "index.ts",
        "filename": "index",
        "isEntry": true,
        "parent": {
          "items": [Circular],
          "name": "a",
          "parent": {
            "items": [],
            "name": "root",
            "path": "/",
            "relativePath": "/",
            "relativePathName": "",
            "subDirectory": {
              "a": [Circular],
            },
          },
          "path": "/a/",
          "relativePath": "a/",
          "relativePathName": "a",
          "subDirectory": {},
        },
        "path": "/a/index.ts",
      },
    ]
  `)
  // get items from file, should return []
  expect(builder.getItems('/a/index.ts')).toEqual([])
  // get non-exist path, should return []
  expect(builder.getItems('/a/b/')).toEqual([])
  // get root
  expect(builder.getItems(ROOT_NAME)).toMatchInlineSnapshot(`[]`)
})

it('custom separator', () => {
  const builder = new PathTreeBuilder({ sep: '$' })
  builder.addPath('$a$')
  builder.addPath('$a$b$')
  builder.addPath('$a$b$index.ts')
  builder.addPath('$a$b$foo.ts')
  expect(builder.tree).toMatchSnapshot('custom-separator add path')
  builder.removePath('$a$b$')
  builder.removePath('$a$b$foo.ts')
  builder.removePath('$a$b$baz.ts') // remove non-exist path
  expect(builder.tree).toMatchSnapshot('custom-separator remove path')
})

it('get data', () => {
  const builder = new PathTreeBuilder({
    getData: node => ({ data: node.ext }),
  })
  builder.addPath('/a/index.ts')
  // override default
  builder.addPath('/a/foo.ts', { data: 'hello world' })
  const node = builder.getNode('/a/index.ts', true)
  expect(node?.data).toStrictEqual({
    data: 'ts',
  })
  expect(builder.getNode('/a/foo.ts', true)?.data).toStrictEqual({
    data: 'hello world',
  })
  // get non-exist path, should return null
  expect(builder.getNode('/a/b/index.ts')).toBeNull()
})
