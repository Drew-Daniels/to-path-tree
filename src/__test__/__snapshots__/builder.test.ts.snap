// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom separator > custom-separator add path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a$",
      "relativePath": "a$",
      "relativePathName": "a",
      "subDirectory": {
        "b": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "$a$b$index.ts",
            },
            {
              "data": undefined,
              "ext": "ts",
              "file": "foo.ts",
              "filename": "foo",
              "isEntry": false,
              "parent": [Circular],
              "path": "$a$b$foo.ts",
            },
          ],
          "name": "b",
          "parent": [Circular],
          "path": "/a$b$",
          "relativePath": "b$",
          "relativePathName": "b",
          "subDirectory": {},
        },
      },
    },
  },
}
`;

exports[`custom separator > custom-separator remove path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a$",
      "relativePath": "a$",
      "relativePathName": "a",
      "subDirectory": {},
    },
  },
}
`;

exports[`simple builder > simple-builder add path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {
        "b": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "/a/b/index.ts",
            },
            {
              "data": undefined,
              "ext": "ts",
              "file": "foo.ts",
              "filename": "foo",
              "isEntry": false,
              "parent": [Circular],
              "path": "/a/b/foo.ts",
            },
          ],
          "name": "b",
          "parent": [Circular],
          "path": "/a/b/",
          "relativePath": "b/",
          "relativePathName": "b",
          "subDirectory": {},
        },
      },
    },
  },
}
`;

exports[`simple builder > simple-builder remove path 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {},
    },
  },
}
`;
